

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Ficheiros de texto &mdash; Bq computacional (FCUL) May 2018 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/my_style.css" type="text/css" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Funções e módulos" href="funcs_modules.html" />
    <link rel="prev" title="Strings" href="strings.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Bq computacional (FCUL)
          

          
          </a>

          
            
            
              <div class="version">
                5.2018
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="basics.html">Conceitos básicos</a></li>
<li class="toctree-l1"><a class="reference internal" href="collections_iterations.html">Coleções e iterações</a></li>
<li class="toctree-l1"><a class="reference internal" href="more_on_lists_dicts.html">Listas e dicionários</a></li>
<li class="toctree-l1"><a class="reference internal" href="strings.html"><em>Strings</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Ficheiros de texto</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#leitura">Leitura</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#read-com-open-e-close-explicitos"><code class="docutils literal"><span class="pre">.read()</span></code>, com <code class="docutils literal"><span class="pre">open()</span></code> e <code class="docutils literal"><span class="pre">close()</span></code> explícitos</a></li>
<li class="toctree-l3"><a class="reference internal" href="#read-dentro-do-bloco-de-um-comando-with"><code class="docutils literal"><span class="pre">.read()</span></code>, dentro do bloco de um comando <code class="docutils literal"><span class="pre">with</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#readlines"><code class="docutils literal"><span class="pre">.readlines()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#iteracao-de-ficheiros-com-for">Iteração de ficheiros com <code class="docutils literal"><span class="pre">for</span></code>.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#exemplo-extracao-de-informacao-de-um-ficheiro-fasta-multiplo">Exemplo: Extração de informação de um ficheiro FASTA múltiplo.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#escrita">Escrita</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#funcao-print-para-ficheiros">Função <code class="docutils literal"><span class="pre">print()</span></code> para ficheiros</a></li>
<li class="toctree-l3"><a class="reference internal" href="#funcao-write">Função <code class="docutils literal"><span class="pre">.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#exemplo-extracao-de-informacao-de-ficheiros-de-resultados-de-metabolomica">Exemplo: Extração de informação de ficheiros de resultados de metabolómica.</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#exploracao-do-formato">Exploração do formato</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#correspondencia-compostos-vias">Correspondência compostos - vias</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#utilizacao-da-informacao">Utilização da informação</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#informacao-obtida-por-acesso-a-internet-modulo-requests">Informação obtida por acesso à Internet: (módulo <code class="docutils literal"><span class="pre">requests</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="funcs_modules.html">Funções e módulos</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_scientific_modules.html">Python “científico”</a></li>
<li class="toctree-l1"><a class="reference internal" href="11_scipy_regression.html">Exemplo do uso de <code class="docutils literal"><span class="pre">numpy</span></code> e <code class="docutils literal"><span class="pre">scipy</span></code>: regressão linear.</a></li>
<li class="toctree-l1"><a class="reference internal" href="12_scipy_pH.html">Example: Simulation of the acid-base changes in an amino-acid solution</a></li>
<li class="toctree-l1"><a class="reference internal" href="13_algoritmos.html">Algoritmos numéricos</a></li>
<li class="toctree-l1"><a class="reference internal" href="14_pandas.html">Python “científico”: biblioteca <em>pandas</em></a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_software.html">Instalação de software</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Bq computacional (FCUL)</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Ficheiros de texto</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="ficheiros-de-texto">
<h1>Ficheiros de texto<a class="headerlink" href="#ficheiros-de-texto" title="Permalink to this headline">¶</a></h1>
<div class="section" id="leitura">
<h2>Leitura<a class="headerlink" href="#leitura" title="Permalink to this headline">¶</a></h2>
<p>O modelo mais simples é ler todo o conteúdo de um ficheiro para uma
<em>string</em>:</p>
<div class="figure">
<img alt="" src="_images/fichs.png" />
</div>
<p>A leitura de um ficheiro segundo este modelo é feito através da função
<code class="docutils literal"><span class="pre">.read()</span></code>.</p>
<p>Mas o processo é um pouco mais complicado do que o uso simples de uma
função.</p>
<p>O acesso (programático) a um ficheiro existente num computador requer
que num programa se indique que esse acesso vai começar, a <em>abertura</em> de
um ficheiro e que o acesso vai terminar, o <em>fecho</em> de um ficheiro.</p>
<div class="section" id="read-com-open-e-close-explicitos">
<h3><code class="docutils literal"><span class="pre">.read()</span></code>, com <code class="docutils literal"><span class="pre">open()</span></code> e <code class="docutils literal"><span class="pre">close()</span></code> explícitos<a class="headerlink" href="#read-com-open-e-close-explicitos" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">a</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;A sequência, em FASTA é&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>&lt;class &#39;str&#39;&gt;
A sequência, em FASTA é
&gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN
DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS
KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG
DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA
DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI
GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL
GDNAVFAGENFHHGDKL
</pre></div>
</div>
</div>
<div class="section" id="read-dentro-do-bloco-de-um-comando-with">
<h3><code class="docutils literal"><span class="pre">.read()</span></code>, dentro do bloco de um comando <code class="docutils literal"><span class="pre">with</span></code><a class="headerlink" href="#read-dentro-do-bloco-de-um-comando-with" title="Permalink to this headline">¶</a></h3>
<p>Numa versão mais “moderna” podemos abrir e <strong>automaticamente fechar</strong> o
ficheiro é utilizar o comando <code class="docutils literal"><span class="pre">with</span></code>:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;A sequência, em FASTA é&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>A sequência, em FASTA é
&gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN
DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS
KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG
DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA
DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI
GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL
GDNAVFAGENFHHGDKL
</pre></div>
</div>
<p>O comando <code class="docutils literal"><span class="pre">with</span></code> faz o ficheiro permanecer <em>aberto</em> até ao fim do
“bloco”, (também aqui) indicado pelo alinhamento mais à direita de um ou
mais comandos a seguir à linha em que se encontra o <code class="docutils literal"><span class="pre">with</span></code>. Quando
termina o bloco o ficheiro é fechado sem usar a função <code class="docutils literal"><span class="pre">close()</span></code>.</p>
<p>Além de <code class="docutils literal"><span class="pre">read()</span></code>, em que todo o conteúdo de um ficheiro é lido para
uma <em>string</em>, existem outras maneiras de ler um ficheiro.</p>
</div>
<div class="section" id="readlines">
<h3><code class="docutils literal"><span class="pre">.readlines()</span></code><a class="headerlink" href="#readlines" title="Permalink to this headline">¶</a></h3>
<p>A função <code class="docutils literal"><span class="pre">readlines()</span></code> lê e separa <strong>as linhas</strong> de um ficheiro para
uma lista:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>[&#39;&gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]\n&#39;, &#39;MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN\n&#39;, &#39;DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS\n&#39;, &#39;KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG\n&#39;, &#39;DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA\n&#39;, &#39;DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI\n&#39;, &#39;GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL\n&#39;, &#39;GDNAVFAGENFHHGDKL\n&#39;, &#39;\n&#39;]
</pre></div>
</div>
<p>O que são os <code class="docutils literal"><span class="pre">\n</span></code> no fim das <em>strings</em>?</p>
<p><strong>Numa string,</strong> <code class="docutils literal"><span class="pre">\n</span></code> <strong>indica a mudança de linha</strong>. (Conta como
apenas <strong>1</strong> caractere).</p>
<p>Neste caso eles aparecem porque no ficheiro original há mudanças de
linha.</p>
<p>Muitas vezes, é necessário elimina-los. Para isso podemos usar a função
<code class="docutils literal"><span class="pre">.strip()</span></code>:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

<span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">linha</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">linha</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>[&#39;&gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]&#39;, &#39;MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN&#39;, &#39;DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS&#39;, &#39;KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG&#39;, &#39;DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA&#39;, &#39;DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI&#39;, &#39;GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL&#39;, &#39;GDNAVFAGENFHHGDKL&#39;, &#39;&#39;]
</pre></div>
</div>
<p>Ou, de uma forma sucinta, usando uma lista em compreensão:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">linha</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">linha</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>[&#39;&gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]&#39;, &#39;MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN&#39;, &#39;DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS&#39;, &#39;KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG&#39;, &#39;DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA&#39;, &#39;DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI&#39;, &#39;GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL&#39;, &#39;GDNAVFAGENFHHGDKL&#39;, &#39;&#39;]
</pre></div>
</div>
<p>Com ficheiros muito grandes, a leitura pelas funções <code class="docutils literal"><span class="pre">.read()</span></code> e
<code class="docutils literal"><span class="pre">.readlines()</span></code> pode esgotar a memória de um computador e “congelar” um
programa.</p>
<p>Existe uma terceira maneira de ler um ficheiro (que não traz problemas
com ficheiros grandes):</p>
</div>
<div class="section" id="iteracao-de-ficheiros-com-for">
<h3>Iteração de ficheiros com <code class="docutils literal"><span class="pre">for</span></code>.<a class="headerlink" href="#iteracao-de-ficheiros-com-for" title="Permalink to this headline">¶</a></h3>
<p><strong>A iteração de um ficheiro “percorre” as linhas do ficheiro</strong></p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">linha</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
        <span class="n">linha</span> <span class="o">=</span> <span class="n">linha</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Linha:&#39;</span><span class="p">,</span> <span class="n">linha</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>Linha: &gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
Linha: MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN
Linha: DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS
Linha: KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG
Linha: DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA
Linha: DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI
Linha: GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL
Linha: GDNAVFAGENFHHGDKL
Linha:
</pre></div>
</div>
<p>Podemos até usar a função <code class="docutils literal"><span class="pre">enumerate()</span></code> com um ficheiro. São gerados
os pares de valores</p>
<p><code class="docutils literal"><span class="pre">(num</span> <span class="pre">linha,</span> <span class="pre">linha)</span></code>.</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">linha</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="n">linha</span> <span class="o">=</span> <span class="n">linha</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;linha&#39;</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">linha</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>linha 0 : &gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
linha 1 : MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVN
linha 2 : DVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKS
linha 3 : KTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVG
linha 4 : DEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLA
linha 5 : DLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQI
linha 6 : GTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL
linha 7 : GDNAVFAGENFHHGDKL
linha 8 :
</pre></div>
</div>
<p><strong>Problema: ler uma ficheiro FASTA e separar o cabeçalho da sequência em
duas strings (juntando toda a sequência numa só string)</strong></p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span>

<span class="n">header</span> <span class="o">=</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c1"># usamos um slice de uma lista de 1 até ao fim</span>
<span class="n">outras</span> <span class="o">=</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
<span class="c1"># e a funçao .join() com separador vazio para</span>
<span class="c1"># juntá-las</span>
<span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outras</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cabeçalho:&quot;</span><span class="p">,</span> <span class="n">header</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sequência, com&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="s1">&#39;aminoácidos:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>cabeçalho: &gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
sequência, com 437 aminoácidos:
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVNDVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKSKTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVGDEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLADLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQIGTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEELGDNAVFAGENFHHGDKL
</pre></div>
</div>
<p>Às vezes os ficheiros não têm cabeçalho! É melhor testar se a primeira
linha começa por “&gt;” !</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;eno1.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="k">if</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cabeçalho:&quot;</span><span class="p">,</span> <span class="n">header</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sequência, com&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="s1">&#39;aminoácidos:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>cabeçalho: &gt;gi|398366315|ref|NP_011770.3| Eno1p [Saccharomyces cerevisiae S288c]
sequência, com 437 aminoácidos:
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVNDVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKSKTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVGDEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLADLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQIGTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEELGDNAVFAGENFHHGDKL
</pre></div>
</div>
<p>As linhas em branco podem por vezes causar alguns problemas. Mas é fácil
“ignora-las”.</p>
<p>Vamos supor que o ficheiro <strong>gre3.txt</strong> tem o seguinte conteúdo:</p>
<hr class="docutils" />
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">sp</span><span class="o">|</span><span class="n">P38715</span><span class="o">|</span><span class="n">GRE3_YEAST</span> <span class="n">NADPH</span><span class="o">-</span><span class="n">dependent</span> <span class="n">aldose</span> <span class="n">reductase</span> <span class="n">GRE3</span> <span class="n">OS</span><span class="o">=</span><span class="n">Saccharomyces</span> <span class="n">cerevisiae</span> <span class="p">(</span><span class="n">strain</span> <span class="n">ATCC</span> <span class="mi">204508</span> <span class="o">/</span> <span class="n">S288c</span><span class="p">)</span> <span class="n">GN</span><span class="o">=</span><span class="n">GRE3</span> <span class="n">PE</span><span class="o">=</span><span class="mi">1</span> <span class="n">SV</span><span class="o">=</span><span class="mi">1</span>

<span class="n">MSSLVTLNNGLKMPLVGLGCWKIDKKVCANQIYEAIKLGYRLFDGACDYGNEKEVGEGIR</span>
<span class="n">KAISEGLVSRKDIFVVSKLWNNFHHPDHVKLALKKTLSDMGLDYLDLYYIHFPIAFKYVP</span>
<span class="n">FEEKYPPGFYTGADDEKKGHITEAHVPIIDTYRALEECVDEGLIKSIGVSNFQGSLIQDL</span>
<span class="n">LRGCRIKPVALQIEHHPYLTQEHLVEFCKLHDIQVVAYSSFGPQSFIEMDLQLAKTTPTL</span>
<span class="n">FENDVIKKVSQNHPGSTTSQVLLRWATQRGIAVIPKSSKKERLLGNLEIEKKFTLTEQEL</span>
<span class="n">KDISALNANIRFNDPWTWLDGKFPTFA</span>
</pre></div>
</div>
<hr class="docutils" />
<p>Como separar o cabeçalho da sequência?</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;gre3.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">linhas</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="k">if</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cabeçalho:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sequência, com&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="s1">&#39;aminoácidos:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>cabeçalho:
&gt;sp|P38715|GRE3_YEAST NADPH-dependent aldose reductase GRE3 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=GRE3 PE=1 SV=1
sequência, com 327 aminoácidos:
MSSLVTLNNGLKMPLVGLGCWKIDKKVCANQIYEAIKLGYRLFDGACDYGNEKEVGEGIRKAISEGLVSRKDIFVVSKLWNNFHHPDHVKLALKKTLSDMGLDYLDLYYIHFPIAFKYVPFEEKYPPGFYTGADDEKKGHITEAHVPIIDTYRALEECVDEGLIKSIGVSNFQGSLIQDLLRGCRIKPVALQIEHHPYLTQEHLVEFCKLHDIQVVAYSSFGPQSFIEMDLQLAKTTPTLFENDVIKKVSQNHPGSTTSQVLLRWATQRGIAVIPKSSKKERLLGNLEIEKKFTLTEQELKDISALNANIRFNDPWTWLDGKFPTFA
</pre></div>
</div>
</div>
</div>
<div class="section" id="exemplo-extracao-de-informacao-de-um-ficheiro-fasta-multiplo">
<h2>Exemplo: Extração de informação de um ficheiro FASTA múltiplo.<a class="headerlink" href="#exemplo-extracao-de-informacao-de-um-ficheiro-fasta-multiplo" title="Permalink to this headline">¶</a></h2>
<p><strong>Problema: extraír os cabeçalhos e as sequências de um ficheiro FASTA
múltiplo. Mostrar o comprimento das proteínas e o número de triptofanos
(W)</strong></p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;proteins.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">tudo</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">prots</span> <span class="o">=</span> <span class="n">tudo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>0
1121
1151
374
551
549
551
351
556
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;proteins.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">tudo</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">prots</span> <span class="o">=</span> <span class="n">tudo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
<span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">[:</span><span class="mi">30</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>1121
sp|P16862|PFKA2_YEAST ATP-depe
1151
sp|P16861|PFKA1_YEAST ATP-depe
374
sp|P00950|PMG1_YEAST Phosphogl
551
sp|P00924|ENO1_YEAST Enolase 1
549
sp|P30575|ENO1_CANAL Enolase 1
551
sp|P00925|ENO2_YEAST Enolase 2
351
sp|P32626|ENOPH_YEAST Enolase-
556
sp|P40370|ENO11_SCHPO Enolase
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;proteins.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">tudo</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">prots</span> <span class="o">=</span> <span class="n">tudo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
<span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">p</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)]</span>
    <span class="n">headers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>

<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">headers</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>sp|P16862|PFKA2_YEAST ATP-dependent 6-phosphofructokinase subunit beta OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=PFK2 PE=1 SV=4
sp|P16861|PFKA1_YEAST ATP-dependent 6-phosphofructokinase subunit alpha OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=PFK1 PE=1 SV=1
sp|P00950|PMG1_YEAST Phosphoglycerate mutase 1 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=GPM1 PE=1 SV=3
sp|P00924|ENO1_YEAST Enolase 1 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=ENO1 PE=1 SV=3
sp|P30575|ENO1_CANAL Enolase 1 OS=Candida albicans (strain SC5314 / ATCC MYA-2876) GN=ENO1 PE=2 SV=1
sp|P00925|ENO2_YEAST Enolase 2 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=ENO2 PE=1 SV=2
sp|P32626|ENOPH_YEAST Enolase-phosphatase E1 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) GN=UTR4 PE=1 SV=2
sp|P40370|ENO11_SCHPO Enolase 1-1 OS=Schizosaccharomyces pombe (strain 972 / ATCC 24843) GN=eno101 PE=1 SV=2
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;proteins.fasta&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">tudo</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">prots</span> <span class="o">=</span> <span class="n">tudo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
<span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">prots</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">p</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)]</span>
    <span class="n">headers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>

<span class="n">ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">headers</span><span class="p">:</span>
    <span class="n">separados</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)</span>
    <span class="n">ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">separados</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">s</span>  <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span> <span class="n">seqs</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;tem&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">),</span> <span class="s1">&#39;aminoácidos,&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;W&#39;</span><span class="p">),</span> <span class="s1">&#39;são triptofanos&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>P16862 tem 959 aminoácidos, 10 são triptofanos
P16861 tem 987 aminoácidos, 12 são triptofanos
P00950 tem 247 aminoácidos, 5 são triptofanos
P00924 tem 437 aminoácidos, 5 são triptofanos
P30575 tem 440 aminoácidos, 5 são triptofanos
P00925 tem 437 aminoácidos, 5 são triptofanos
P32626 tem 227 aminoácidos, 1 são triptofanos
P40370 tem 439 aminoácidos, 7 são triptofanos
</pre></div>
</div>
</div>
<div class="section" id="escrita">
<h2>Escrita<a class="headerlink" href="#escrita" title="Permalink to this headline">¶</a></h2>
<div class="section" id="funcao-print-para-ficheiros">
<h3>Função <code class="docutils literal"><span class="pre">print()</span></code> para ficheiros<a class="headerlink" href="#funcao-print-para-ficheiros" title="Permalink to this headline">¶</a></h3>
<p>Basta abrir o ficheiro em <em>modo de escrita</em> usando o argumento <code class="docutils literal"><span class="pre">w</span></code> na
função <code class="docutils literal"><span class="pre">open()</span></code>. Depois, modificar a função <code class="docutils literal"><span class="pre">print()</span></code>, com o
argumento <code class="docutils literal"><span class="pre">file</span></code>, indicando que o resultado da escrita deve ser
<em>enviado</em> para o ficheiro.</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1, 2, 3, experiência, som, som&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">a</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">30</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="o">**</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">a</span><span class="p">)</span>
</pre></div>
</div>
<p>Aparentemente não aconteceu nada, mas um ficheiro novo foi criado</p>
<p>Vamos ler o ficheiro:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>1, 2, 3, experiência, som, som
0 0.0
1 1.0
2 1.4142135623730951
3 1.7320508075688772
4 2.0
5 2.23606797749979
6 2.449489742783178
7 2.6457513110645907
8 2.8284271247461903
9 3.0
10 3.1622776601683795
11 3.3166247903554
12 3.4641016151377544
13 3.605551275463989
14 3.7416573867739413
15 3.872983346207417
16 4.0
17 4.123105625617661
18 4.242640687119285
19 4.358898943540674
20 4.47213595499958
21 4.58257569495584
22 4.69041575982343
23 4.795831523312719
24 4.898979485566356
25 5.0
26 5.0990195135927845
27 5.196152422706632
28 5.291502622129181
29 5.385164807134504
</pre></div>
</div>
</div>
<div class="section" id="funcao-write">
<h3>Função <code class="docutils literal"><span class="pre">.write()</span></code><a class="headerlink" href="#funcao-write" title="Permalink to this headline">¶</a></h3>
<p>Também existe a função <code class="docutils literal"><span class="pre">.write()</span></code> que funciona como o contrário de
<code class="docutils literal"><span class="pre">.read()</span></code>:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">tudo</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">Um texto que ocupa</span>
<span class="s2">1 linha</span>
<span class="s2">2 linhas</span>
<span class="s2">3 linhas</span>
<span class="s2">&quot;&quot;&quot;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp2.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">a</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tudo</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp2.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>Um texto que ocupa
1 linha
2 linhas
3 linhas
</pre></div>
</div>
<p><strong>Problema: ler uma ficheiro com dados numéricos e converter o ponto
decimal em vírgula decimal</strong></p>
<p>No ficheiro <code class="docutils literal"><span class="pre">exp.txt</span></code>, recentemente criado, podemos, de uma form
sucinta, passar os <code class="docutils literal"><span class="pre">.</span></code> a <code class="docutils literal"><span class="pre">,</span></code> ?</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">tudo</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;,&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">a</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tudo</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;exp.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>1, 2, 3, experiência, som, som
0 0,0
1 1,0
2 1,4142135623730951
3 1,7320508075688772
4 2,0
5 2,23606797749979
6 2,449489742783178
7 2,6457513110645907
8 2,8284271247461903
9 3,0
10 3,1622776601683795
11 3,3166247903554
12 3,4641016151377544
13 3,605551275463989
14 3,7416573867739413
15 3,872983346207417
16 4,0
17 4,123105625617661
18 4,242640687119285
19 4,358898943540674
20 4,47213595499958
21 4,58257569495584
22 4,69041575982343
23 4,795831523312719
24 4,898979485566356
25 5,0
26 5,0990195135927845
27 5,196152422706632
28 5,291502622129181
29 5,385164807134504
</pre></div>
</div>
</div>
</div>
<div class="section" id="exemplo-extracao-de-informacao-de-ficheiros-de-resultados-de-metabolomica">
<h2>Exemplo: Extração de informação de ficheiros de resultados de metabolómica.<a class="headerlink" href="#exemplo-extracao-de-informacao-de-ficheiros-de-resultados-de-metabolomica" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://www.masstrix.org">MassTRIX</a>, (<em>Mass TRanslator into
Pathways</em>) [1] é um serviço online de tratamento de dados de
metabolómica.</p>
<p>A funcionalidade primária é a identificação de compostos a partir de
listas de massas e intensidades obtidas por análise de amostras
biológics por Espectrometria de Massa.</p>
<p>O resultado da identificação é disponibilizado em vários ficheiros de
texto. Num dos formatos, cada linha do ficheiro diz respeito a um pico
de massa e apresenta, de entre outros, os compostos identificados com
aquela massa, bem como as anotações das vias celulares em que esses
compostos podem estar envolvidos.</p>
<p>Pretende-se ilustrar o uso programático da leitura de ficheiros e as
operações com <em>strings</em> com um exemplo da <strong>extração da informação
contida num desses ficheiros</strong>.</p>
<p>[1] K. Suhre and P. Schmitt-Kopplin (2008) MassTRIX: Mass TRanslator
Into Pathways, <em>Nucleic Acids Research</em>, <strong>36</strong>, Web Server issue,
W481-W484.</p>
<div class="section" id="exploracao-do-formato">
<h3>Exploração do formato<a class="headerlink" href="#exploracao-do-formato" title="Permalink to this headline">¶</a></h3>
<p>Vamos ler o ficheiro <code class="docutils literal"><span class="pre">masses.annotated.reformat.tsv</span></code>, separar todas as
linhas para uma lista e mostrar a primeira e a última:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;FIRST line ----------------------------&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">all_lines</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;LAST line -----------------------------&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>FIRST line ----------------------------
154.97517   7.25775e+06     120.005768420091        4       154.975098039829#154.975098039829#154.975274805989      0.464333550973771#0.464333550973771#-0.676276005999922  C00988#HMDB00816#C02287 C2H5O6P#C2H5O6P#C3H4O5  2-Phosphoglycolate;Phosphoglycolic acid ([M-H]-)#Phosphoglycolic acid (see KEGG C00988); 2-phosphonooxyacetic acid [carboxylic acid] ([M-H]-)#Hydroxymalonate;Tartronic acid;Hydroxymalonic acid;2-Hydroxymalonate;2-Hydroxymalonic acid;2-Tartronic acid ([M+Cl35]-)                                                                                                   ko00630;ko01100#null#null       ;Glyoxylate and dicarboxylate metabolism;Metabolic pathways#null#null   null#null#null
LAST line -----------------------------
raw_mass    peak_height     corrected_mass  npossible       KEGG_mass       ppm     KEGG_cid        KEGG_formula    KEGG_name       uniqueID        C13     O18     N15     S34     Mg25    Mg26    Fe54    Fe57    Ca44    Cl37    K41     KEGG Pathways   KEGG Pathways descriptions      Compound in Organism(X)
</pre></div>
</div>
<p>Nas linhas deste ficheiro, os vários campos com informação estão
separados por <strong>tabs</strong> (o caractere <code class="docutils literal"><span class="pre">\t</span></code>).</p>
<p>A última linha tem como informação os nomes de cada um destes campos
(<code class="docutils literal"><span class="pre">raw_mass</span> <span class="pre">peak_height</span></code> etc)</p>
<p>Vamos dividir a linha 0 em várias partes, pelo separador <code class="docutils literal"><span class="pre">\t</span></code>. As
partes obtidas são os vários campos de informação reltiva a um pico de
MS.</p>
<p>Já agora, vamos obter os nomes de cada campo, fazendo o mesmo à última
linha:</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">headers</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>raw_mass
peak_height
corrected_mass
npossible
KEGG_mass
ppm
KEGG_cid
KEGG_formula
KEGG_name
uniqueID
C13
O18
N15
S34
Mg25
Mg26
Fe54
Fe57
Ca44
Cl37
K41
KEGG Pathways
KEGG Pathways descriptions
Compound in Organism(X)
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">line0</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">info</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">headers</span><span class="p">,</span> <span class="n">line0</span><span class="p">))</span>

<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">headers</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>raw_mass : 154.97517
peak_height : 7.25775e+06
corrected_mass : 120.005768420091
npossible : 4
KEGG_mass : 154.975098039829#154.975098039829#154.975274805989
ppm : 0.464333550973771#0.464333550973771#-0.676276005999922
KEGG_cid : C00988#HMDB00816#C02287
KEGG_formula : C2H5O6P#C2H5O6P#C3H4O5
KEGG_name : 2-Phosphoglycolate;Phosphoglycolic acid ([M-H]-)#Phosphoglycolic acid (see KEGG C00988); 2-phosphonooxyacetic acid [carboxylic acid] ([M-H]-)#Hydroxymalonate;Tartronic acid;Hydroxymalonic acid;2-Hydroxymalonate;2-Hydroxymalonic acid;2-Tartronic acid ([M+Cl35]-)
uniqueID :
C13 :
O18 :
N15 :
S34 :
Mg25 :
Mg26 :
Fe54 :
Fe57 :
Ca44 :
Cl37 :
K41 :
KEGG Pathways : ko00630;ko01100#null#null
KEGG Pathways descriptions : ;Glyoxylate and dicarboxylate metabolism;Metabolic pathways#null#null
Compound in Organism(X) : null#null#null
</pre></div>
</div>
<p>Vamos extraír da linha 0</p>
<ul class="simple">
<li>a massa do pico “<em>raw mass</em>”, (campo 0)</li>
<li>a intensidade do pico, (campo 1)</li>
<li>os IDs dos compostos, (campo 6)</li>
<li>os nomes dos compostos (campo 8)</li>
<li>os IDs das vias (campo 21)</li>
<li>as descrições das vias (campo 22)</li>
</ul>
<p>Havendo vários compostos possíveis em cada pico, é usado como separador
o caractere <code class="docutils literal"><span class="pre">#</span></code>.</p>
<p>Podemos já separar a informação por composto.</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="n">use_only</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;raw_mass&#39;</span><span class="p">,</span> <span class="s1">&#39;peak_height&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>
<span class="n">needs_split</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">line0</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">h</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">headers</span><span class="p">,</span> <span class="n">line0</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
        <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">needs_split</span><span class="p">:</span>
    <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>raw_mass : 154.97517
peak_height : 7.25775e+06
KEGG_cid : [&#39;C00988&#39;, &#39;HMDB00816&#39;, &#39;C02287&#39;]
KEGG_name : [&#39;2-Phosphoglycolate;Phosphoglycolic acid ([M-H]-)&#39;, &#39;Phosphoglycolic acid (see KEGG C00988); 2-phosphonooxyacetic acid [carboxylic acid] ([M-H]-)&#39;, &#39;Hydroxymalonate;Tartronic acid;Hydroxymalonic acid;2-Hydroxymalonate;2-Hydroxymalonic acid;2-Tartronic acid ([M+Cl35]-)&#39;]
KEGG Pathways : [&#39;ko00630;ko01100&#39;, &#39;null&#39;, &#39;null&#39;]
KEGG Pathways descriptions : [&#39;;Glyoxylate and dicarboxylate metabolism;Metabolic pathways&#39;, &#39;null&#39;, &#39;null&#39;]
</pre></div>
</div>
<p>Quanto à informação relativa às vias em que cada composto pode estar
envolvido, podemos reparar que:</p>
<ol class="arabic simple">
<li>Um composto pde ter várias vias, separadas por <code class="docutils literal"><span class="pre">;</span></code>.</li>
<li>Um composto pode não ter nenhuma via. neste caso, aparece a anotação
“null”.</li>
</ol>
<p>Finalmente, vamos transformar a informação relativa às vias (quer os IDs
quer as descrições) em <strong>listas</strong>.</p>
<p>Repare-se que ainda são <em>strings</em> e que usam como separador o <code class="docutils literal"><span class="pre">;</span></code> para
delimitar várias vias.</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="n">use_only</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;raw_mass&#39;</span><span class="p">,</span> <span class="s1">&#39;peak_height&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>
<span class="n">needs_split</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>
<span class="n">needs_more_split</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">line0</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">h</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">headers</span><span class="p">,</span> <span class="n">line0</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
        <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">needs_split</span><span class="p">:</span>
    <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">needs_more_split</span><span class="p">:</span>
    <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span>

<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>raw_mass : 154.97517
peak_height : 7.25775e+06
KEGG_cid : [&#39;C00988&#39;, &#39;HMDB00816&#39;, &#39;C02287&#39;]
KEGG_name : [&#39;2-Phosphoglycolate;Phosphoglycolic acid ([M-H]-)&#39;, &#39;Phosphoglycolic acid (see KEGG C00988); 2-phosphonooxyacetic acid [carboxylic acid] ([M-H]-)&#39;, &#39;Hydroxymalonate;Tartronic acid;Hydroxymalonic acid;2-Hydroxymalonate;2-Hydroxymalonic acid;2-Tartronic acid ([M+Cl35]-)&#39;]
KEGG Pathways : [[&#39;ko00630&#39;, &#39;ko01100&#39;], [&#39;null&#39;], [&#39;null&#39;]]
KEGG Pathways descriptions : [[&#39;&#39;, &#39;Glyoxylate and dicarboxylate metabolism&#39;, &#39;Metabolic pathways&#39;], [&#39;null&#39;], [&#39;null&#39;]]
</pre></div>
</div>
<p>Agora <strong>tudo junto, aplicando ao ficheiro inteiro</strong>. Para controlo,
podemos contar os compostos obtidos.</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;masses.annotated.reformat.tsv&#39;</span>
<span class="n">use_only</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;raw_mass&#39;</span><span class="p">,</span> <span class="s1">&#39;peak_height&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>
<span class="n">needs_split</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG_cid&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG_name&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>
<span class="n">needs_more_split</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">,</span> <span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">all_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>

<span class="n">headers</span> <span class="o">=</span> <span class="n">all_lines</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">peaks</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">all_lines</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
    <span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">line_parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">h</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">headers</span><span class="p">,</span> <span class="n">line_parts</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
            <span class="n">info</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">needs_split</span><span class="p">:</span>
        <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">needs_more_split</span><span class="p">:</span>
        <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span>

    <span class="n">peaks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;São&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">peaks</span><span class="p">),</span> <span class="s1">&#39;massas&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">---- Massa 0 -----&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">use_only</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">peaks</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">h</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>São 482 massas

---- Massa 0 -----
raw_mass : 154.97517
peak_height : 7.25775e+06
KEGG_cid : [&#39;C00988&#39;, &#39;HMDB00816&#39;, &#39;C02287&#39;]
KEGG_name : [&#39;2-Phosphoglycolate;Phosphoglycolic acid ([M-H]-)&#39;, &#39;Phosphoglycolic acid (see KEGG C00988); 2-phosphonooxyacetic acid [carboxylic acid] ([M-H]-)&#39;, &#39;Hydroxymalonate;Tartronic acid;Hydroxymalonic acid;2-Hydroxymalonate;2-Hydroxymalonic acid;2-Tartronic acid ([M+Cl35]-)&#39;]
KEGG Pathways : [[&#39;ko00630&#39;, &#39;ko01100&#39;], [&#39;null&#39;], [&#39;null&#39;]]
KEGG Pathways descriptions : [[&#39;Glyoxylate and dicarboxylate metabolism&#39;, &#39;Metabolic pathways&#39;], [&#39;null&#39;], [&#39;null&#39;]]
</pre></div>
</div>
<div class="section" id="correspondencia-compostos-vias">
<h4>Correspondência compostos - vias<a class="headerlink" href="#correspondencia-compostos-vias" title="Permalink to this headline">¶</a></h4>
<p>Agora com esta <strong>lista de dicionários</strong> chamada <code class="docutils literal"><span class="pre">peaks</span></code> disponível
podemos centrar a informação em torrno dos compostos associados a vias.</p>
<p>Para isso vamos criar dois dicionários:</p>
<ul class="simple">
<li>um chamado <code class="docutils literal"><span class="pre">pathways</span></code> que associa cada <strong>Id de um composto</strong> a uma
<strong>lista de Ids de vias</strong> e</li>
<li>outro chamado <code class="docutils literal"><span class="pre">descriptions</span></code>, que associa cada Id de via à sua
descrição.</li>
</ul>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">paths</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">descriptions</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="s1">&#39;KEGG_cid&#39;</span><span class="p">],</span> <span class="n">k</span><span class="p">[</span><span class="s1">&#39;KEGG Pathways&#39;</span><span class="p">],</span> <span class="n">k</span><span class="p">[</span><span class="s1">&#39;KEGG Pathways descriptions&#39;</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;null&#39;</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>
        <span class="k">for</span> <span class="n">pId</span><span class="p">,</span> <span class="n">desc</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">d</span><span class="p">):</span>
            <span class="n">descriptions</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span> <span class="o">=</span> <span class="n">desc</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;São&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">paths</span><span class="p">),</span> <span class="s1">&#39;compostos com anotações de vias&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">---------Alguns compostos:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">paths</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
        <span class="k">break</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="s1">&#39;--&gt;&#39;</span><span class="p">,</span> <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">---------Alguns compostos:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">paths</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
        <span class="k">break</span>
    <span class="n">p_desc</span> <span class="o">=</span> <span class="p">[</span><span class="n">descriptions</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">]]</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="s1">&#39;--&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39; AND &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">p_desc</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>São 327 compostos com anotações de vias

---------Alguns compostos:

C00988 --&gt; [&#39;ko00630&#39;, &#39;ko01100&#39;]
C16652 --&gt; [&#39;ko00982&#39;]
C16655 --&gt; [&#39;ko00982&#39;]
C01088 --&gt; [&#39;ko00770&#39;]
C01989 --&gt; [&#39;ko00630&#39;]
C02488 --&gt; [&#39;ko00620&#39;]
C02991 --&gt; [&#39;ko00051&#39;]
C03652 --&gt; [&#39;ko00760&#39;]
C03979 --&gt; [&#39;ko00051&#39;]
C06159 --&gt; [&#39;ko00051&#39;]
C16390 --&gt; [&#39;ko00760&#39;]

---------Alguns compostos:

C00988 --&gt; Glyoxylate and dicarboxylate metabolism AND Metabolic pathways
C16652 --&gt; Drug metabolism - cytochrome P450
C16655 --&gt; Drug metabolism - cytochrome P450
C01088 --&gt; Pantothenate and CoA biosynthesis
C01989 --&gt; Glyoxylate and dicarboxylate metabolism
C02488 --&gt; Pyruvate metabolism
C02991 --&gt; Fructose and mannose metabolism
C03652 --&gt; Nicotinate and nicotinamide metabolism
C03979 --&gt; Fructose and mannose metabolism
C06159 --&gt; Fructose and mannose metabolism
C16390 --&gt; Nicotinate and nicotinamide metabolism
</pre></div>
</div>
</div>
</div>
<div class="section" id="utilizacao-da-informacao">
<h3>Utilização da informação<a class="headerlink" href="#utilizacao-da-informacao" title="Permalink to this headline">¶</a></h3>
<p>Agora com estes dois dicionários podemos responder a várias questões:</p>
<p>Exemplo: Como obter uma <strong>lista</strong> com nomes das vias, mas sem
repetições?</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">]:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">descriptions</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
            <span class="n">names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

<span class="c1"># AS primeiras 20 vias:</span>
<span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">[:</span><span class="mi">21</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>Glyoxylate and dicarboxylate metabolism
Metabolic pathways
Drug metabolism - cytochrome P450
Pantothenate and CoA biosynthesis
Pyruvate metabolism
Fructose and mannose metabolism
Nicotinate and nicotinamide metabolism
Phenylalanine metabolism
Phenylalanine, tyrosine and tryptophan biosynthesis
Phenylpropanoid biosynthesis
Tropane, piperidine and pyridine alkaloid biosynthesis
Glucosinolate biosynthesis
Aminoacyl-tRNA biosynthesis
Biosynthesis of phenylpropanoids
Biosynthesis of alkaloids derived from shikimate pathway
Biosynthesis of alkaloids derived from ornithine, lysine and nicotinic acid
Biosynthesis of plant hormones
ABC transporters
Biosynthesis of plant secondary metabolites
Alanine, aspartate and glutamate metabolism
Tetracycline biosynthesis
</pre></div>
</div>
<p>Exemplo: Como obter um <strong>dicionário</strong> com os <strong>Ids das vias como
chaves</strong> e o <strong>número de vezes que aparecem como valores</strong>?</p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">counts</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;São&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">counts</span><span class="p">),</span> <span class="s1">&#39;vias&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">---------Algumas contagens:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pId</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">counts</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">pId</span><span class="p">],</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">pId</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">descriptions</span><span class="p">[</span><span class="n">pId</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>São 150 vias

---------Algumas contagens:

8    ko00630         Glyoxylate and dicarboxylate metabolism
113          ko01100         Metabolic pathways
5    ko00982         Drug metabolism - cytochrome P450
4    ko00770         Pantothenate and CoA biosynthesis
4    ko00620         Pyruvate metabolism
17   ko00051         Fructose and mannose metabolism
6    ko00760         Nicotinate and nicotinamide metabolism
4    ko00360         Phenylalanine metabolism
6    ko00400         Phenylalanine, tyrosine and tryptophan biosynthesis
2    ko00940         Phenylpropanoid biosynthesis
</pre></div>
</div>
<p>Uma vez que <code class="docutils literal"><span class="pre">counts</span></code> é um dicionário, não se aplica a noção de ordem e
é evidente que as vias não estão ordenadas segundo as contagens de
compostos.</p>
<p>Podemos obter as vias por ordem decrescente de compostos?</p>
<p>Para, por exemplo, obter <strong>as 20 vias mais abundantes</strong> em compostos?</p>
<p>Uma vez que os dicionários não estão associados a uma “ordenação”, temos
de trabalhar com listas.</p>
<p>Estratégia:</p>
<ul class="simple">
<li>Criar uma lista com os pares (contagens, Id da via)</li>
<li>Ordenar a lista</li>
</ul>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">counts_list</span> <span class="o">=</span> <span class="p">[(</span><span class="n">counts</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">]</span>

<span class="c1">#Controlo: 5 primeiros elementos, lista desordenada:</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">counts_list</span><span class="p">[:</span><span class="mi">5</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>(8, &#39;ko00630&#39;)
(113, &#39;ko01100&#39;)
(5, &#39;ko00982&#39;)
(4, &#39;ko00770&#39;)
(4, &#39;ko00620&#39;)
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">counts_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># reverse=True indica que a ordenação é por ordem decrescente</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;As 20 vias com mais compostos associados:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">counts_list</span><span class="p">[:</span><span class="mi">20</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">descriptions</span><span class="p">[</span><span class="n">pId</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>As 20 vias com mais compostos associados:

113 : Metabolic pathways
24 : Biosynthesis of plant secondary metabolites
21 : Galactose metabolism
20 : alpha-Linolenic acid metabolism
18 : Biosynthesis of unsaturated fatty acids
17 : Biosynthesis of terpenoids and steroids
17 : Linoleic acid metabolism
17 : Starch and sucrose metabolism
17 : Ascorbate and aldarate metabolism
17 : Fructose and mannose metabolism
16 : Phosphotransferase system (PTS)
15 : Diterpenoid biosynthesis
15 : Steroid biosynthesis
14 : Biosynthesis of plant hormones
14 : Glycolysis / Gluconeogenesis
13 : ABC transporters
13 : Biosynthesis of alkaloids derived from shikimate pathway
13 : Biosynthesis of phenylpropanoids
13 : Amino sugar and nucleotide sugar metabolism
13 : Pentose and glucuronate interconversions
</pre></div>
</div>
<p>Como exemplo final, escrever um ficheiro que sumariza esta informação.</p>
<p>Problema: Escrever um ficheiro, chamado <code class="docutils literal"><span class="pre">pathways.txt</span></code> com vários
campos, separados por <code class="docutils literal"><span class="pre">\t</span></code> e <strong>uma via por linha</strong>.</p>
<p>As vias devem estar por ordem decrescente de ocorrência</p>
<p>Os campos são:</p>
<ul class="simple">
<li>O ID da via</li>
<li>A descrição da via</li>
<li>O número de ocorrências</li>
<li>Os Ids dos compostos associados à via, separados por <code class="docutils literal"><span class="pre">;</span></code></li>
</ul>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">file_name</span> <span class="o">=</span> <span class="s1">&#39;pathways.txt&#39;</span>

<span class="c1"># associação entre ids de vias e lista de compostos</span>
<span class="n">compounds</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">[</span><span class="n">c</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">pId</span> <span class="ow">in</span> <span class="n">compounds</span><span class="p">:</span>
            <span class="n">compounds</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">compounds</span><span class="p">[</span><span class="n">pId</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span><span class="p">]</span>

<span class="c1"># contagens de coorrências (número de compostos)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">Id</span> <span class="ow">in</span> <span class="n">compounds</span><span class="p">:</span>
    <span class="n">counts</span><span class="p">[</span><span class="n">Id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">compounds</span><span class="p">[</span><span class="n">Id</span><span class="p">])</span>

<span class="c1"># ordenar as contagens</span>
<span class="n">counts_list</span> <span class="o">=</span> <span class="p">[(</span><span class="n">counts</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">]</span>
<span class="n">counts_list</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># escrever o ficheiro com a informação</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">Id</span> <span class="ow">in</span> <span class="n">counts_list</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">Id</span><span class="p">,</span> <span class="n">descriptions</span><span class="p">[</span><span class="n">Id</span><span class="p">],</span> <span class="n">c</span><span class="p">,</span> <span class="n">compounds</span><span class="p">[</span><span class="n">Id</span><span class="p">],</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="c1"># verificar se correu bem...</span>
<span class="n">file_name</span> <span class="o">=</span> <span class="s1">&#39;pathways.txt&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">a</span><span class="p">:</span>
    <span class="n">linhas</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">14</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>ko00010     Glycolysis / Gluconeogenesis    14      [&#39;C00111&#39;, &#39;C00118&#39;, &#39;C00031&#39;, &#39;C00221&#39;, &#39;C00267&#39;, &#39;C00631&#39;, &#39;C00197&#39;, &#39;C00103&#39;, &#39;C00668&#39;, &#39;C01172&#39;, &#39;C05345&#39;, &#39;C00236&#39;, &#39;C01159&#39;, &#39;C16255&#39;]
</pre></div>
</div>
</div>
</div>
<div class="section" id="informacao-obtida-por-acesso-a-internet-modulo-requests">
<h2>Informação obtida por acesso à Internet: (módulo <code class="docutils literal"><span class="pre">requests</span></code>)<a class="headerlink" href="#informacao-obtida-por-acesso-a-internet-modulo-requests" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;http://www.uniprot.org/uniprot/P00924.fasta&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>&gt;sp|P00924|ENO1_YEAST Enolase 1 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 GN=ENO1 PE=1 SV=3
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGK
GVLHAVKNVNDVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASR
AAAAEKNVPLYKHLADLSKSKTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTF
AEALRIGSEVYHNLKSLTKKRYGASAGNVGDEGGVAPNIQTAEEALDLIVDAIKAAGHDG
KIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLADLYHSLMKRYPIVSIEDPFA
EDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQIGTLSESIKAA
QDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEEL
GDNAVFAGENFHHGDKL
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="n">linhas</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">if</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
    <span class="n">cab</span> <span class="o">=</span> <span class="n">linhas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">cab</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">linhas</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cabeçalho: &quot;</span><span class="p">,</span> <span class="n">cab</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;sequência:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>cabeçalho:  &gt;sp|P00924|ENO1_YEAST Enolase 1 OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 GN=ENO1 PE=1 SV=3
sequência:
MAVSKVYARSVYDSRGNPTVEVELTTEKGVFRSIVPSGASTGVHEALEMRDGDKSKWMGKGVLHAVKNVNDVIAPAFVKANIDVKDQKAVDDFLISLDGTANKSKLGANAILGVSLAASRAAAAEKNVPLYKHLADLSKSKTSPYVLPVPFLNVLNGGSHAGGALALQEFMIAPTGAKTFAEALRIGSEVYHNLKSLTKKRYGASAGNVGDEGGVAPNIQTAEEALDLIVDAIKAAGHDGKIKIGLDCASSEFFKDGKYDLDFKNPNSDKSKWLTGPQLADLYHSLMKRYPIVSIEDPFAEDDWEAWSHFFKTAGIQIVADDLTVTNPKRIATAIEKKAADALLLKVNQIGTLSESIKAAQDSFAAGWGVMVSHRSGETEDTFIADLVVGLRTGQIKTGAPARSERLAKLNQLLRIEEELGDNAVFAGENFHHGDKL
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;http://www.uniprot.org/uniprot/P00924.txt&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>ID   ENO1_YEAST              Reviewed;         437 AA.
AC   P00924; D6VV34; P99013;
DT   21-JUL-1986, integrated into UniProtKB/Swiss-Prot.
DT   05-OCT-2010, sequence version 3.
DT   28-MAR-2018, entry version 203.
DE   RecName: Full=Enolase 1;
DE            EC=4.2.1.11;
DE   AltName: Full=2-phospho-D-glycerate hydro-lyase 1;
DE   AltName: Full=2-phosphoglycerate dehydratase 1;
GN   Name=ENO1; Synonyms=ENOA, HSP48; OrderedLocusNames=YGR254W;
GN   ORFNames=G9160;
OS   Saccharomyces cerevisiae (strain ATCC 204508 / S288c) (Baker&#39;s yeast).
OC   Eukaryota; Fungi; Dikarya; Ascomycota; Saccharomycotina;
OC   Saccharomycetes; Saccharomycetales; Saccharomycetaceae; Saccharomyces.
OX   NCBI_TaxID=559292;
RN   [1]
RP   NUCLEOTIDE SEQUENCE [GENOMIC DNA].
RX   PubMed=6256394;
RA   Holland M.J., Holland J.P., Thill G.P., Jackson K.A.;
RT   &quot;The primary structures of two yeast enolase genes. Homology between
RT   the 5&#39; noncoding flanking regions of yeast enolase and glyceraldehyde-
RT   3-phosphate dehydrogenase genes.&quot;;
RL   J. Biol. Chem. 256:1385-1395(1981).
RN   [2]
RP   NUCLEOTIDE SEQUENCE [GENOMIC DNA].
RC   STRAIN=ATCC 204508 / S288c;
RX   PubMed=9133741;
RX   DOI=10.1002/(SICI)1097-0061(19970330)13:4&lt;369::AID-YEA81&gt;3.0.CO;2-V;
RA   Mazzoni C., Ruzzi M., Rinaldi T., Solinas F., Montebove F.,
RA   Frontali L.;
RT   &quot;Sequence analysis of a 10.5 kb DNA fragment from the yeast chromosome
RT   VII reveals the presence of three new open reading frames and of a
RT   tRNAThr gene.&quot;;
RL   Yeast 13:369-372(1997).
RN   [3]
RP   NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].
RC   STRAIN=ATCC 204508 / S288c;
RX   PubMed=9169869;
RA   Tettelin H., Agostoni-Carbone M.L., Albermann K., Albers M.,
RA   Arroyo J., Backes U., Barreiros T., Bertani I., Bjourson A.J.,
RA   Brueckner M., Bruschi C.V., Carignani G., Castagnoli L., Cerdan E.,
RA   Clemente M.L., Coblenz A., Coglievina M., Coissac E., Defoor E.,
RA   Del Bino S., Delius H., Delneri D., de Wergifosse P., Dujon B.,
RA   Durand P., Entian K.-D., Eraso P., Escribano V., Fabiani L.,
RA   Fartmann B., Feroli F., Feuermann M., Frontali L., Garcia-Gonzalez M.,
RA   Garcia-Saez M.I., Goffeau A., Guerreiro P., Hani J., Hansen M.,
RA   Hebling U., Hernandez K., Heumann K., Hilger F., Hofmann B.,
RA   Indge K.J., James C.M., Klima R., Koetter P., Kramer B., Kramer W.,
RA   Lauquin G., Leuther H., Louis E.J., Maillier E., Marconi A.,
RA   Martegani E., Mazon M.J., Mazzoni C., McReynolds A.D.K.,
RA   Melchioretto P., Mewes H.-W., Minenkova O., Mueller-Auer S.,
RA   Nawrocki A., Netter P., Neu R., Nombela C., Oliver S.G., Panzeri L.,
RA   Paoluzi S., Plevani P., Portetelle D., Portillo F., Potier S.,
RA   Purnelle B., Rieger M., Riles L., Rinaldi T., Robben J.,
RA   Rodrigues-Pousada C., Rodriguez-Belmonte E., Rodriguez-Torres A.M.,
RA   Rose M., Ruzzi M., Saliola M., Sanchez-Perez M., Schaefer B.,
RA   Schaefer M., Scharfe M., Schmidheini T., Schreer A., Skala J.,
RA   Souciet J.-L., Steensma H.Y., Talla E., Thierry A., Vandenbol M.,
RA   van der Aart Q.J.M., Van Dyck L., Vanoni M., Verhasselt P., Voet M.,
RA   Volckaert G., Wambutt R., Watson M.D., Weber N., Wedler E., Wedler H.,
RA   Wipfli P., Wolf K., Wright L.F., Zaccaria P., Zimmermann M.,
RA   Zollner A., Kleine K.;
RT   &quot;The nucleotide sequence of Saccharomyces cerevisiae chromosome VII.&quot;;
RL   Nature 387:81-84(1997).
RN   [4]
RP   GENOME REANNOTATION.
RC   STRAIN=ATCC 204508 / S288c;
RX   PubMed=24374639; DOI=10.1534/g3.113.008995;
RA   Engel S.R., Dietrich F.S., Fisk D.G., Binkley G., Balakrishnan R.,
RA   Costanzo M.C., Dwight S.S., Hitz B.C., Karra K., Nash R.S., Weng S.,
RA   Wong E.D., Lloyd P., Skrzypek M.S., Miyasato S.R., Simison M.,
RA   Cherry J.M.;
RT   &quot;The reference genome sequence of Saccharomyces cerevisiae: Then and
RT   now.&quot;;
RL   G3 (Bethesda) 4:389-398(2014).
RN   [5]
RP   PROTEIN SEQUENCE OF 2-437.
RX   PubMed=7005235;
RA   Chin C.C.Q., Brewer J.M., Wold F.;
RT   &quot;The amino acid sequence of yeast enolase.&quot;;
RL   J. Biol. Chem. 256:1377-1384(1981).
RN   [6]
RP   PROTEIN SEQUENCE OF 2-12.
RC   STRAIN=ATCC 26786 / X2180-1A;
RA   Sanchez J.-C., Golaz O., Schaller D., Morch F., Frutiger S.,
RA   Hughes G.J., Appel R.D., Deshusses J., Hochstrasser D.F.;
RL   Submitted (AUG-1995) to UniProtKB.
RN   [7]
RP   PROTEIN SEQUENCE OF 30-47.
RC   STRAIN=ATCC 204508 / S288c;
RX   PubMed=7895733; DOI=10.1002/elps.11501501210;
RA   Garrels J.I., Futcher B., Kobayashi R., Latter G.I., Schwender B.,
RA   Volpe T., Warner J.R., McLaughlin C.S.;
RT   &quot;Protein identifications for a Saccharomyces cerevisiae protein
RT   database.&quot;;
RL   Electrophoresis 15:1466-1486(1994).
RN   [8]
RP   PROTEIN SEQUENCE OF 69-79.
RC   STRAIN=ATCC 38531 / Y41;
RX   PubMed=7737086; DOI=10.1002/elps.1150160124;
RA   Norbeck J., Blomberg A.;
RT   &quot;Gene linkage of two-dimensional polyacrylamide gel electrophoresis
RT   resolved proteins from isogene families in Saccharomyces cerevisiae by
RT   microsequencing of in-gel trypsin generated peptides.&quot;;
RL   Electrophoresis 16:149-156(1995).
RN   [9]
RP   MUTAGENESIS OF LYS-346.
RX   PubMed=8634301; DOI=10.1021/bi952186y;
RA   Poyner R.R., Laughlin L.T., Sowa G.A., Reed G.H.;
RT   &quot;Toward identification of acid/base catalysts in the active site of
RT   enolase: comparison of the properties of K345A, E168Q, and E211Q
RT   variants.&quot;;
RL   Biochemistry 35:1692-1699(1996).
RN   [10]
RP   MUTAGENESIS OF HIS-160.
RX   PubMed=11027610; DOI=10.1006/bbrc.2000.3618;
RA   Brewer J.M., Holland M.J., Lebioda L.;
RT   &quot;The H159A mutant of yeast enolase 1 has significant activity.&quot;;
RL   Biochem. Biophys. Res. Commun. 276:1199-1202(2000).
RN   [11]
RP   SUBCELLULAR LOCATION.
RX   PubMed=11502169; DOI=10.1021/bi010277r;
RA   Grandier-Vazeille X., Bathany K., Chaignepain S., Camougrand N.,
RA   Manon S., Schmitter J.-M.;
RT   &quot;Yeast mitochondrial dehydrogenases are associated in a supramolecular
RT   complex.&quot;;
RL   Biochemistry 40:9758-9769(2001).
RN   [12]
RP   MUTAGENESIS OF HIS-160 AND ASN-208.
RX   PubMed=13678299; DOI=10.1023/A:1025390123761;
RA   Brewer J.M., Glover C.V., Holland M.J., Lebioda L.;
RT   &quot;Enzymatic function of loop movement in enolase: preparation and some
RT   properties of H159N, H159A, H159F, and N207A enolases.&quot;;
RL   J. Protein Chem. 22:353-361(2003).
RN   [13]
RP   LEVEL OF PROTEIN EXPRESSION [LARGE SCALE ANALYSIS].
RX   PubMed=14562106; DOI=10.1038/nature02046;
RA   Ghaemmaghami S., Huh W.-K., Bower K., Howson R.W., Belle A.,
RA   Dephoure N., O&#39;Shea E.K., Weissman J.S.;
RT   &quot;Global analysis of protein expression in yeast.&quot;;
RL   Nature 425:737-741(2003).
RN   [14]
RP   PHOSPHORYLATION [LARGE SCALE ANALYSIS] AT SER-119, AND IDENTIFICATION
RP   BY MASS SPECTROMETRY [LARGE SCALE ANALYSIS].
RX   PubMed=17287358; DOI=10.1073/pnas.0607084104;
RA   Chi A., Huttenhower C., Geer L.Y., Coon J.J., Syka J.E.P., Bai D.L.,
RA   Shabanowitz J., Burke D.J., Troyanskaya O.G., Hunt D.F.;
RT   &quot;Analysis of phosphorylation sites on proteins from Saccharomyces
RT   cerevisiae by electron transfer dissociation (ETD) mass
RT   spectrometry.&quot;;
RL   Proc. Natl. Acad. Sci. U.S.A. 104:2193-2198(2007).
RN   [15]
RP   UBIQUITINATION [LARGE SCALE ANALYSIS] AT LYS-358, AND IDENTIFICATION
RP   BY MASS SPECTROMETRY [LARGE SCALE ANALYSIS].
RX   PubMed=22106047; DOI=10.1002/pmic.201100166;
RA   Starita L.M., Lo R.S., Eng J.K., von Haller P.D., Fields S.;
RT   &quot;Sites of ubiquitin attachment in Saccharomyces cerevisiae.&quot;;
RL   Proteomics 12:236-240(2012).
RN   [16]
RP   X-RAY CRYSTALLOGRAPHY (2.25 ANGSTROMS).
RX   PubMed=3374614; DOI=10.1038/333683a0;
RA   Lebioda L., Stec B.;
RT   &quot;Crystal structure of enolase indicates that enolase and pyruvate
RT   kinase evolved from a common ancestor.&quot;;
RL   Nature 333:683-686(1988).
RN   [17]
RP   X-RAY CRYSTALLOGRAPHY (2.25 ANGSTROMS).
RX   PubMed=2645275;
RA   Lebioda L., Stec B., Brewer J.M.;
RT   &quot;The structure of yeast enolase at 2.25-A resolution. An 8-fold beta +
RT   alpha-barrel with a novel beta beta alpha alpha (beta alpha)6
RT   topology.&quot;;
RL   J. Biol. Chem. 264:3685-3693(1989).
RN   [18]
RP   X-RAY CRYSTALLOGRAPHY (2.25 ANGSTROMS).
RX   PubMed=2405163; DOI=10.1016/0022-2836(90)90023-F;
RA   Stec B., Lebioda L.;
RT   &quot;Refined structure of yeast apo-enolase at 2.25-A resolution.&quot;;
RL   J. Mol. Biol. 211:235-248(1990).
RN   [19]
RP   X-RAY CRYSTALLOGRAPHY (1.8 ANGSTROMS) IN COMPLEX WITH SUBSTRATE AND
RP   MAGNESIUM IONS.
RX   PubMed=8605183; DOI=10.1021/bi952859c;
RA   Larsen T.M., Wedekind J.E., Rayment I., Reed G.H.;
RT   &quot;A carboxylate oxygen of the substrate bridges the magnesium ions at
RT   the active site of enolase: structure of the yeast enzyme complexed
RT   with the equilibrium mixture of 2-phosphoglycerate and
RT   phosphoenolpyruvate at 1.8-A resolution.&quot;;
RL   Biochemistry 35:4349-4358(1996).
RN   [20]
RP   X-RAY CRYSTALLOGRAPHY (2.0 ANGSTROMS) IN COMPLEX WITH SUBSTRATE.
RX   PubMed=9376357; DOI=10.1021/bi9712450;
RA   Zhang E., Brewer J.M., Minor W., Carreira L.A., Lebioda L.;
RT   &quot;Mechanism of enolase: the crystal structure of asymmetric dimer
RT   enolase-2-phospho-D-glycerate/enolase-phosphoenolpyruvate at 2.0-A
RT   resolution.&quot;;
RL   Biochemistry 36:12526-12534(1997).
RN   [21]
RP   X-RAY CRYSTALLOGRAPHY (2.1 ANGSTROMS) OF MUTANT ALA-40 IN COMPLEX WITH
RP   MAGNESIUM IONS AND SUBSTRATE ANALOG.
RX   PubMed=12054465; DOI=10.1016/S0003-9861(02)00024-3;
RA   Poyner R.R., Larsen T.M., Wong S.-W., Reed G.H.;
RT   &quot;Functional and structural changes due to a serine to alanine mutation
RT   in the active-site flap of enolase.&quot;;
RL   Arch. Biochem. Biophys. 401:155-163(2002).
RN   [22]
RP   X-RAY CRYSTALLOGRAPHY (1.8 ANGSTROMS) OF MUTANT GLN-212 AND MUTANT
RP   GLN-169.
RX   PubMed=12846578; DOI=10.1021/bi0346345;
RA   Sims P.A., Larsen T.M., Poyner R.R., Cleland W.W., Reed G.H.;
RT   &quot;Reverse protonation is the key to general acid-base catalysis in
RT   enolase.&quot;;
RL   Biochemistry 42:8298-8306(2003).
CC   -!- CATALYTIC ACTIVITY: 2-phospho-D-glycerate = phosphoenolpyruvate +
CC       H(2)O.
CC   -!- COFACTOR:
CC       Name=Mg(2+); Xref=ChEBI:CHEBI:18420;
CC       Note=Mg(2+) is required for catalysis and for stabilizing the
CC       dimer.;
CC   -!- PATHWAY: Carbohydrate degradation; glycolysis; pyruvate from D-
CC       glyceraldehyde 3-phosphate: step 4/5.
CC   -!- SUBUNIT: Homodimer. {ECO:0000269|PubMed:12054465,
CC       ECO:0000269|PubMed:8605183, ECO:0000269|PubMed:9376357}.
CC   -!- INTERACTION:
CC       P11484:SSB1; NbExp=3; IntAct=EBI-6468, EBI-8627;
CC   -!- SUBCELLULAR LOCATION: Cytoplasm {ECO:0000269|PubMed:11502169}.
CC   -!- MISCELLANEOUS: Present with 76700 molecules/cell in log phase SD
CC       medium. {ECO:0000269|PubMed:14562106}.
CC   -!- SIMILARITY: Belongs to the enolase family. {ECO:0000305}.
CC   -----------------------------------------------------------------------
CC   Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms
CC   Distributed under the Creative Commons Attribution-NoDerivs License
CC   -----------------------------------------------------------------------
DR   EMBL; J01322; AAA88712.1; -; Genomic_DNA.
DR   EMBL; X99228; CAA67616.1; -; Genomic_DNA.
DR   EMBL; Z73039; CAA97283.1; -; Genomic_DNA.
DR   EMBL; BK006941; DAA08345.1; -; Genomic_DNA.
DR   PIR; S64586; NOBY.
DR   RefSeq; NP_011770.3; NM_001181383.3.
DR   PDB; 1EBG; X-ray; 2.10 A; A/B=2-437.
DR   PDB; 1EBH; X-ray; 1.90 A; A/B=2-437.
DR   PDB; 1ELS; X-ray; 2.40 A; A=2-437.
DR   PDB; 1L8P; X-ray; 2.10 A; A/B/C/D=2-437.
DR   PDB; 1NEL; X-ray; 2.60 A; A=2-437.
DR   PDB; 1ONE; X-ray; 1.80 A; A/B=2-437.
DR   PDB; 1P43; X-ray; 1.80 A; A/B=2-437.
DR   PDB; 1P48; X-ray; 2.00 A; A/B=2-437.
DR   PDB; 2AL1; X-ray; 1.50 A; A/B=2-437.
DR   PDB; 2AL2; X-ray; 1.85 A; A/B=2-437.
DR   PDB; 2ONE; X-ray; 2.00 A; A/B=2-437.
DR   PDB; 2XGZ; X-ray; 1.80 A; A/B=2-437.
DR   PDB; 2XH0; X-ray; 1.70 A; A/B/C/D=2-437.
DR   PDB; 2XH2; X-ray; 1.80 A; A/B/C/D=2-437.
DR   PDB; 2XH4; X-ray; 1.70 A; A/B/C/D=2-437.
DR   PDB; 2XH7; X-ray; 1.80 A; A/B=2-437.
DR   PDB; 3ENL; X-ray; 2.25 A; A=2-437.
DR   PDB; 4ENL; X-ray; 1.90 A; A=2-437.
DR   PDB; 5ENL; X-ray; 2.20 A; A=2-437.
DR   PDB; 6ENL; X-ray; 2.20 A; A=2-437.
DR   PDB; 7ENL; X-ray; 2.20 A; A=2-437.
DR   PDBsum; 1EBG; -.
DR   PDBsum; 1EBH; -.
DR   PDBsum; 1ELS; -.
DR   PDBsum; 1L8P; -.
DR   PDBsum; 1NEL; -.
DR   PDBsum; 1ONE; -.
DR   PDBsum; 1P43; -.
DR   PDBsum; 1P48; -.
DR   PDBsum; 2AL1; -.
DR   PDBsum; 2AL2; -.
DR   PDBsum; 2ONE; -.
DR   PDBsum; 2XGZ; -.
DR   PDBsum; 2XH0; -.
DR   PDBsum; 2XH2; -.
DR   PDBsum; 2XH4; -.
DR   PDBsum; 2XH7; -.
DR   PDBsum; 3ENL; -.
DR   PDBsum; 4ENL; -.
DR   PDBsum; 5ENL; -.
DR   PDBsum; 6ENL; -.
DR   PDBsum; 7ENL; -.
DR   ProteinModelPortal; P00924; -.
DR   SMR; P00924; -.
DR   BioGrid; 33505; 132.
DR   DIP; DIP-5561N; -.
DR   IntAct; P00924; 105.
DR   MINT; P00924; -.
DR   STRING; 4932.YGR254W; -.
DR   Allergome; 786; Sac c Enolase.
DR   CarbonylDB; P00924; -.
DR   iPTMnet; P00924; -.
DR   COMPLUYEAST-2DPAGE; P00924; -.
DR   SWISS-2DPAGE; P00924; -.
DR   UCD-2DPAGE; P00924; -.
DR   MaxQB; P00924; -.
DR   PaxDb; P00924; -.
DR   PRIDE; P00924; -.
DR   TopDownProteomics; P00924; -.
DR   EnsemblFungi; YGR254W; YGR254W; YGR254W.
DR   GeneID; 853169; -.
DR   KEGG; sce:YGR254W; -.
DR   EuPathDB; FungiDB:YGR254W; -.
DR   SGD; S000003486; ENO1.
DR   GeneTree; ENSGT00910000144064; -.
DR   HOGENOM; HOG000072174; -.
DR   InParanoid; P00924; -.
DR   KO; K01689; -.
DR   OMA; EFMIIPV; -.
DR   OrthoDB; EOG092C2W5X; -.
DR   BioCyc; YEAST:YGR254W-MONOMER; -.
DR   BRENDA; 4.2.1.11; 984.
DR   SABIO-RK; P00924; -.
DR   UniPathway; UPA00109; UER00187.
DR   EvolutionaryTrace; P00924; -.
DR   PRO; PR:P00924; -.
DR   Proteomes; UP000002311; Chromosome VII.
DR   GO; GO:0000324; C:fungal-type vacuole; IDA:SGD.
DR   GO; GO:0005739; C:mitochondrion; IDA:SGD.
DR   GO; GO:0000015; C:phosphopyruvate hydratase complex; IDA:SGD.
DR   GO; GO:0000287; F:magnesium ion binding; IEA:InterPro.
DR   GO; GO:0004634; F:phosphopyruvate hydratase activity; IMP:SGD.
DR   GO; GO:0006094; P:gluconeogenesis; IEP:SGD.
DR   GO; GO:0006096; P:glycolytic process; IMP:SGD.
DR   GO; GO:0032889; P:regulation of vacuole fusion, non-autophagic; IDA:SGD.
DR   CDD; cd03313; enolase; 1.
DR   Gene3D; 3.20.20.120; -; 1.
DR   Gene3D; 3.30.390.10; -; 1.
DR   HAMAP; MF_00318; Enolase; 1.
DR   InterPro; IPR000941; Enolase.
DR   InterPro; IPR036849; Enolase-like_C_sf.
DR   InterPro; IPR029017; Enolase-like_N.
DR   InterPro; IPR034390; Enolase-like_superfamily.
DR   InterPro; IPR020810; Enolase_C.
DR   InterPro; IPR020809; Enolase_CS.
DR   InterPro; IPR020811; Enolase_N.
DR   PANTHER; PTHR11902; PTHR11902; 1.
DR   Pfam; PF00113; Enolase_C; 1.
DR   Pfam; PF03952; Enolase_N; 1.
DR   PIRSF; PIRSF001400; Enolase; 1.
DR   PRINTS; PR00148; ENOLASE.
DR   SFLD; SFLDG00178; enolase; 1.
DR   SFLD; SFLDS00001; Enolase; 1.
DR   SMART; SM01192; Enolase_C; 1.
DR   SMART; SM01193; Enolase_N; 1.
DR   SUPFAM; SSF51604; SSF51604; 1.
DR   TIGRFAMs; TIGR01060; eno; 1.
DR   PROSITE; PS00164; ENOLASE; 1.
PE   1: Evidence at protein level;
KW   3D-structure; Complete proteome; Cytoplasm; Direct protein sequencing;
KW   Glycolysis; Isopeptide bond; Lyase; Magnesium; Metal-binding;
KW   Phosphoprotein; Reference proteome; Ubl conjugation.
FT   INIT_MET      1      1       Removed. {ECO:0000269|PubMed:7005235,
FT                                ECO:0000269|Ref.6}.
FT   CHAIN         2    437       Enolase 1.
FT                                /FTId=PRO_0000134062.
FT   REGION      373    376       Substrate binding.
FT   ACT_SITE    212    212       Proton donor. {ECO:0000305}.
FT   ACT_SITE    346    346       Proton acceptor.
FT   METAL       247    247       Magnesium. {ECO:0000269|PubMed:8605183}.
FT   METAL       296    296       Magnesium. {ECO:0000269|PubMed:8605183}.
FT   METAL       321    321       Magnesium. {ECO:0000269|PubMed:8605183}.
FT   BINDING     160    160       Substrate. {ECO:0000269|PubMed:8605183,
FT                                ECO:0000269|PubMed:9376357}.
FT   BINDING     169    169       Substrate. {ECO:0000269|PubMed:8605183,
FT                                ECO:0000269|PubMed:9376357}.
FT   BINDING     296    296       Substrate. {ECO:0000269|PubMed:8605183,
FT                                ECO:0000269|PubMed:9376357}.
FT   BINDING     321    321       Substrate. {ECO:0000269|PubMed:8605183,
FT                                ECO:0000269|PubMed:9376357}.
FT   BINDING     397    397       Substrate. {ECO:0000269|PubMed:8605183,
FT                                ECO:0000269|PubMed:9376357}.
FT   MOD_RES     119    119       Phosphoserine.
FT                                {ECO:0000244|PubMed:17287358}.
FT   MOD_RES     138    138       Phosphoserine.
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   MOD_RES     188    188       Phosphoserine.
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   MOD_RES     313    313       Phosphothreonine.
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   MOD_RES     324    324       Phosphothreonine.
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   CROSSLNK     60     60       Glycyl lysine isopeptide (Lys-Gly)
FT                                (interchain with G-Cter in ubiquitin).
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   CROSSLNK    243    243       Glycyl lysine isopeptide (Lys-Gly)
FT                                (interchain with G-Cter in ubiquitin).
FT                                {ECO:0000250|UniProtKB:P00925}.
FT   CROSSLNK    358    358       Glycyl lysine isopeptide (Lys-Gly)
FT                                (interchain with G-Cter in ubiquitin).
FT                                {ECO:0000244|PubMed:22106047}.
FT   MUTAGEN      40     40       S-&gt;A: Reduces activity by 99.9%.
FT   MUTAGEN     160    160       H-&gt;A,F,N: Reduces activity by 99%.
FT                                {ECO:0000269|PubMed:11027610,
FT                                ECO:0000269|PubMed:13678299}.
FT   MUTAGEN     169    169       E-&gt;Q: Reduces Kcat over 100000-fold.
FT   MUTAGEN     208    208       N-&gt;A: Reduces activity by 44%.
FT                                {ECO:0000269|PubMed:13678299}.
FT   MUTAGEN     212    212       E-&gt;Q: Reduces Kcat over 100000-fold.
FT   MUTAGEN     346    346       K-&gt;A: Reduces Kcat over 100000-fold.
FT                                Abolishes of the proton exchange reaction
FT                                that initiates the enzymatic reaction.
FT                                {ECO:0000269|PubMed:8634301}.
FT   CONFLICT    242    242       I -&gt; V (in Ref. 1; AAA88712).
FT                                {ECO:0000305}.
FT   STRAND        5     12       {ECO:0000244|PDB:2AL1}.
FT   STRAND       18     26       {ECO:0000244|PDB:2AL1}.
FT   STRAND       29     34       {ECO:0000244|PDB:2AL1}.
FT   STRAND       43     45       {ECO:0000244|PDB:1EBH}.
FT   HELIX        57     59       {ECO:0000244|PDB:2AL1}.
FT   HELIX        63     71       {ECO:0000244|PDB:2AL1}.
FT   HELIX        73     80       {ECO:0000244|PDB:2AL1}.
FT   HELIX        87     98       {ECO:0000244|PDB:2AL1}.
FT   STRAND      100    102       {ECO:0000244|PDB:1EBH}.
FT   TURN        104    106       {ECO:0000244|PDB:2AL1}.
FT   HELIX       108    125       {ECO:0000244|PDB:2AL1}.
FT   HELIX       130    138       {ECO:0000244|PDB:2AL1}.
FT   STRAND      145    147       {ECO:0000244|PDB:2AL1}.
FT   STRAND      152    156       {ECO:0000244|PDB:2AL1}.
FT   HELIX       158    160       {ECO:0000244|PDB:2AL1}.
FT   STRAND      161    164       {ECO:0000244|PDB:2AL1}.
FT   STRAND      169    173       {ECO:0000244|PDB:2AL1}.
FT   HELIX       180    202       {ECO:0000244|PDB:2AL1}.
FT   HELIX       204    207       {ECO:0000244|PDB:2AL1}.
FT   STRAND      213    215       {ECO:0000244|PDB:1P48}.
FT   HELIX       222    236       {ECO:0000244|PDB:2AL1}.
FT   TURN        239    241       {ECO:0000244|PDB:2AL1}.
FT   STRAND      243    247       {ECO:0000244|PDB:2AL1}.
FT   HELIX       250    253       {ECO:0000244|PDB:2AL1}.
FT   TURN        261    264       {ECO:0000244|PDB:2AL1}.
FT   HELIX       270    272       {ECO:0000244|PDB:2AL1}.
FT   HELIX       276    289       {ECO:0000244|PDB:2AL1}.
FT   STRAND      292    296       {ECO:0000244|PDB:2AL1}.
FT   HELIX       304    311       {ECO:0000244|PDB:2AL1}.
FT   TURN        312    314       {ECO:0000244|PDB:2AL2}.
FT   STRAND      316    321       {ECO:0000244|PDB:2AL1}.
FT   TURN        322    326       {ECO:0000244|PDB:2AL1}.
FT   HELIX       328    336       {ECO:0000244|PDB:2AL1}.
FT   STRAND      341    345       {ECO:0000244|PDB:2AL1}.
FT   HELIX       347    350       {ECO:0000244|PDB:2AL1}.
FT   HELIX       353    365       {ECO:0000244|PDB:2AL1}.
FT   STRAND      369    373       {ECO:0000244|PDB:2AL1}.
FT   HELIX       383    390       {ECO:0000244|PDB:2AL1}.
FT   STRAND      394    397       {ECO:0000244|PDB:2AL1}.
FT   HELIX       404    420       {ECO:0000244|PDB:2AL1}.
FT   HELIX       421    423       {ECO:0000244|PDB:2AL1}.
FT   STRAND      424    426       {ECO:0000244|PDB:2AL1}.
FT   HELIX       428    430       {ECO:0000244|PDB:2AL1}.
FT   HELIX       434    436       {ECO:0000244|PDB:2AL1}.
SQ   SEQUENCE   437 AA;  46816 MW;  69F45214DBD375BE CRC64;
     MAVSKVYARS VYDSRGNPTV EVELTTEKGV FRSIVPSGAS TGVHEALEMR DGDKSKWMGK
     GVLHAVKNVN DVIAPAFVKA NIDVKDQKAV DDFLISLDGT ANKSKLGANA ILGVSLAASR
     AAAAEKNVPL YKHLADLSKS KTSPYVLPVP FLNVLNGGSH AGGALALQEF MIAPTGAKTF
     AEALRIGSEV YHNLKSLTKK RYGASAGNVG DEGGVAPNIQ TAEEALDLIV DAIKAAGHDG
     KIKIGLDCAS SEFFKDGKYD LDFKNPNSDK SKWLTGPQLA DLYHSLMKRY PIVSIEDPFA
     EDDWEAWSHF FKTAGIQIVA DDLTVTNPKR IATAIEKKAA DALLLKVNQI GTLSESIKAA
     QDSFAAGWGV MVSHRSGETE DTFIADLVVG LRTGQIKTGA PARSERLAKL NQLLRIEEEL
     GDNAVFAGEN FHHGDKL
//
</pre></div>
</div>
<p><strong>Problema:</strong></p>
<ul class="simple">
<li>obter a informação relativa à proteína P00924</li>
<li>filtar a linha começada por <strong>SQ</strong></li>
<li>mostar o numero de aminoácidos e a massa molecular.</li>
</ul>
<p>A informação relativa ao formato desta linha (embora seja evidente
olhando para um exemplo) está descrita na <a class="reference external" href="http://web.expasy.org/docs/userman.html#SQ_line">documentação da
UniProt</a></p>
<p>A linha tem o formato</p>
<p><code class="docutils literal"><span class="pre">SQ</span>&#160;&#160; <span class="pre">SEQUENCE</span> <span class="pre">XXXX</span> <span class="pre">AA;</span> <span class="pre">XXXXX</span> <span class="pre">MW;</span> <span class="pre">XXXXXXXXXXXXXXXX</span> <span class="pre">CRC64;</span></code></p>
<div class="highlight-ipython3"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="n">info</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;http://www.uniprot.org/uniprot/P00924.txt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>

<span class="n">linhas</span> <span class="o">=</span> <span class="n">info</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">sq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">linhas</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;SQ&#39;</span><span class="p">):</span>
        <span class="n">sq</span> <span class="o">=</span> <span class="n">i</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;linha SQ:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sq</span><span class="p">)</span>

<span class="c1"># SQ   SEQUENCE XXXX AA; XXXXX MW; XXXXXXXXXXXXXXXX CRC64;</span>
<span class="n">partes</span> <span class="o">=</span> <span class="n">sq</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">partes</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;aminoácidos&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">partes</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;Da&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>linha SQ:
SQ   SEQUENCE   437 AA;  46816 MW;  69F45214DBD375BE CRC64;
437 aminoácidos
46816 Da
</pre></div>
</div>
<p>Na <a class="reference external" href="http://web.expasy.org/docs/userman.html#FT_keys">documentação da
UniProt</a>, realtiva
às linhas começadas por <code class="docutils literal"><span class="pre">FT</span></code> pode-se ler…</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">INIT_MET</span> <span class="o">-</span> <span class="n">Initiator</span> <span class="n">methionine</span><span class="o">.</span>

<span class="n">This</span> <span class="n">feature</span> <span class="n">key</span> <span class="ow">is</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">a</span> <span class="s1">&#39;1&#39;</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">the</span> <span class="s1">&#39;FROM&#39;</span> <span class="ow">and</span> <span class="s1">&#39;TO&#39;</span> <span class="n">fields</span> <span class="n">to</span> <span class="n">indicate</span> <span class="n">that</span> <span class="n">the</span> <span class="n">initiator</span> <span class="n">methionine</span> <span class="n">has</span> <span class="n">been</span> <span class="n">cleaved</span> <span class="n">off</span><span class="p">:</span>


    <span class="n">FT</span>   <span class="n">INIT_MET</span>      <span class="mi">1</span>      <span class="mi">1</span>       <span class="n">Removed</span><span class="o">.</span>

<span class="n">It</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">used</span> <span class="n">when</span> <span class="n">the</span> <span class="n">initiator</span> <span class="n">methionine</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">cleaved</span> <span class="n">off</span>
</pre></div>
</div>
<p><strong>Problema:</strong></p>
<p>Para as seguintes proteínas,</p>
<p><code class="docutils literal"><span class="pre">Q96UH7,</span> <span class="pre">Q8J0N6,</span> <span class="pre">Q9URB4,</span> <span class="pre">Q9C2U0,</span> <span class="pre">P36580,</span> <span class="pre">P14540</span></code></p>
<p>gerar uma tabela com</p>
<p><code class="docutils literal"><span class="pre">AC</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">MW</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">init</span> <span class="pre">M</span> <span class="pre">cleaved</span></code></p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="funcs_modules.html" class="btn btn-neutral float-right" title="Funções e módulos" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="strings.html" class="btn btn-neutral" title="Strings" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016-2018, António Ferreira.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'May 2018',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          
          SphinxRtdTheme.Navigation.enableSticky();
          
      });
  </script> 

</body>
</html>